src/mlpack/methods/ann/loss_functions/poisson_nll_loss_impl.hpp

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file poisson_nll_loss_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Poisson Negative Log Likelihood class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_IMPL_HPP
+
+
+// In case it hasn't yet been included.
+#include "poisson_nll_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+PoissonNLLLoss<InputDataType, OutputDataType>::PoissonNLLLoss(
+    const bool logInput,
+    const bool full,
+    const double eps,
+    const bool reduction):
+    logInput(logInput),
+    full(full),
+    eps(eps),
+    reduction(reduction)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+double PoissonNLLLoss<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    TargetType& target)
+{
+  InputType loss;
+  loss.set_size(size(input));
+
+  if (logInput)
+    loss = arma::exp(input) - target % input;
+  else
+    loss = input - target % arma::log(input + eps);
+
+  if (full)
+  {
+    loss.elem(arma::find(target > 1)) += target % arma::log(target) - target
+        + arma::log(2 * M_PI * target) / 2;
+  }
+
+  if (reduction)
+    return arma::accu(loss) / input.n_elem;
+  return arma::accu(loss);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType, typename OutputType>
+void PoissonNLLLoss<InputDataType, OutputDataType>::Backward(
+    const InputType& input,
+    const TargetType& target,
+    OutputType& output)
+{
+  output.set_size(size(input));
+  if (logInput)
+    output = (arma::exp(input) - target);
+  else
+    output = (1 - target / (input + eps));
+
+  if (full)
+  {
+    // Here, (1 + 1/2 + 1/3 + ... + 1/n) =~ (log(n + 1) + log(n) + 1)/2.
+    output.elem(arma::find(target > 1)) += (arma::log(target + 1)
+        + arma::log(target) + 1) / 2;
+  }
+
+  if (reduction)
+    output = output / output.n_elem;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename Archive>
+void PoissonNLLLoss<InputDataType, OutputDataType>::serialize(
+    Archive& /* ar */,
+    const unsigned int /* version */)
+{
+  // Nothing to do here.
```

Comment:
In reference to #2329, we should probably serialize the parameters here? I missed this in the review

Reply:
Though I have made the changes but I couldn't understand how serializing parameters in loss functions help?

Reply:
If someone were to choose to serialize a loss function, then it seems intuitive that when they reload the loss function it's behavior remains intact.

Reply:
I mean I understood that but thought that we only need to save weights and the optimizer state. Anyway, this makes sense now.